### ISSUE TITLE
Bombs and rigged cargo chutes that are dropped directly onto ships have no effect

### ISSUE THREAD

**Issue by boredompwndu** [baranidlo/bench#16]
A console message will be generated for bombs, and are not placed on the map, but no ion tokens or damage is assigned.

In the case of rigged cargo chute, no stress is given.

**Comment by Sleet01** [comment]
This is actually a number of different issues:

1) Only Action bombs explode when dropped on a ship; Maneuver bombs only detonate at the end of the activation phase.  All implemented Action bombs are currently detonating correctly when dropped on ships, so that part is working correctly.  **No Work Needed**

2) Connor Nets are properly detonating when dropped on a ship, but when dropped on multiple ships they operate incorrectly:

* Expected: Player is prompted to select a ship to affect; selected ship takes 1 damage, 2 Ion tokens, and the loss of its action this turn (if it has not already activated).  (see [this page](http://xwing-miniatures.wikia.com/wiki/Conner_Net).)  Execution of Activation phase halts until active player chooses affected ship.
* Actual: one ship is automatically selected to suffer the damage and Ion tokens.  Player is given the option to select a ship to suffer the loss of its action, but most likely this will not have any effect because AI units will continue activating while the player is choosing and whatever action it takes will not be rolled back if the player chooses a ship that has already made some action.
* Fix: ideally we would have a way to insert a single, atomic, blocking action into the UI handling stream.  This would allow the game to prompt the user to select a ship, and then apply all Conner Net effects to that ship before continuing the Activation phase.  Unfortunately this is turning out to be a very difficult problem to solve without completely re-writing the game engine (which would be an immense amount of work).   **Needs a concerted effort on the blocking atomic action handler**

3) Rigged Cargo Chute *seems* to work correctly at first glance, because ships that A) have not yet activated and B) get overlapped by the debris, are currently getting the effects applied at the start of their activation.  However this is incorrect because the rules state that the effects should apply immediately and that ships that begin their *maneuver* on an obstacle can ignore that obstacle unless their base or maneuver template overlaps it again.  So we're seeing one incorrect effect application that looks like a correct one that is delayed:

* Expected: When a Rigged Cargo Chute debris field is dropped on a ship, that ship should immediately suffer the Debris Cloud effects (explicitly due to RCC card).  If the affected ship has not yet activated, then during its activation that ship should ignore the Debris Cloud unless its maneuver template passes over the Debris Cloud again, or the ship ends its maneuver on the Debris Cloud again (in which case it suffers the effects a second time).
* Actual: Ships under the RCC-produced Debris Cloud suffer no immediate effects.  When their activation occurs, these ships suffer the Debris Cloud effects even if they should not (e.g. Debris Cloud only covers the back half of the ship and the maneuver template does not pass through it).  End result is similar to Expected but not identical.
* Fix: There are a couple of approaches.  1. Make Rigged Cargo Chute perform a check on all the ships it overlaps before placing its template, and just manually make them take the stress and damage at drop time.  2. make all Obstacles do a check for overlapping ships when they drop, and then have them execute whatever effects they normally produce.  **Either way, though, the Obstacle check code needs to be cleaned up because it looks like ships may be improperly impacted if they are maneuvering *off* of an obstacle.**

tl;dr: most bombs are working correctly, there's a deeper underlying issue with Conner Net vs multiple ships that needs some major work to fix, and both Rigged Cargo Chute and Obstacle checks in general are subtly broken; I only have time to work on one thing so I'll fix RCC and Obstacles this time.

Also, I hate Markdown.

**Comment by Sleet01** [comment]
Okay, Rigged Cargo Chute and the obstacle + maneuver template checks are in the latest Pull Request.
As expected, there was an issue with the template check: the game doesn't use actual templates but instead calculates center-to-center paths based on the projected end-point angle and displacement.  This all works really well until you use the center-to-center path to see if a given maneuver's "template" crosses an obstacle, because about 20mm of each end of that path are either A) not supposed to be checked (lies inside the current ship's base at the current location) or B) already been checked by the maneuver landing-point collision detection code.  So I added a check that aborts the collision detection for those path points which lie inside the ship's base at either its end point or its start point.

The existing code emulates a maneuver template by checking to see if any obstacle's outside edge points fall within 10mm of the maneuver centerline - approximately half the width of the templates to either side - so we may still see some false positives in cases where e.g. the front edge of a ship's base is hanging 9mm off of an obstacle but the rest of the base is on the obstacle.  This *should not* trigger a collision, but with the current code it may if the tolerances are very tight.

The ideal end solution would be to generate a set of accurate template paths in Snap.svg, and use these paths to detect any intersections of the template with any obstacles, but that sort of feature is outside the scope of bug fixes.

**Comment by Sleet01** [comment]
Please use this link to check the bomb / RCC functionality:

http://baranidlo.github.io/bench/main.html?IwZgrANATGBsBcsAMFhihAnEg3MYAHBCBmhlMCvqSLKgCxVjDz0DsdUtWuAZLCGLBUXEQURIUFehAC02HAKEjBwKOLD0iFSPNxKQwtavUSpsGXsWDDKsfE1tosXQt716wtuogpcULQgOGTVUa0FvGykccEFYYDoQG0VMJzACQRR6PBAiWFgnWzBeADMAQwAbAGcAU14AFwAnAFc6gGEKsqqqgEsAY14AfVkABVlgWShhsYmQYYB1XiWkIA

**Comment by Sleet01** [comment]
Rigged Cargo Chute should be working correctly.  Some changes have also been made to maneuver template checking, but more work is needed to make this code 100% TT-identical.

No updates have been made to the obstacle.js code to deal with these issues; there may be some opportunities to fix issues there instead of in individual upgrade cards.

We now have much more knowledge of the doaction / donoaction code, and it should be possible to add a blocking ship-select prompt using code developed for other upgrades (minefield mapper, for example).  So I will add this action to my queue.
